File: app/src/main/java/com/appsys/android/popularmovie/adapter/MovieAdapter.java
package com.appsys.android.popularmovie.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.recyclerview.widget.RecyclerView;

import com.appsys.android.popularmovie.classes.Movie;
import com.shakirfattani.course.movielisting.R;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;

/**
 * Created by shakir on 8/2/2017.
 */

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MovieAdapterViewHolder> {

    private ArrayList<Movie> mMovies;
    private MovieAdapterOnClickHandler mOnClickHandler;

    public MovieAdapter(MovieAdapterOnClickHandler onClick) {
        mOnClickHandler = onClick;
        mMovies = new ArrayList<Movie>();
    }

    public void resetMovieList() {
        mMovies.clear();
        notifyDataSetChanged();
    }

    public ArrayList<Movie> getMoviesData() {
        return mMovies;
    }

    public void setMoviesData(ArrayList<Movie> movieArray) {
        mMovies.addAll(movieArray);
        notifyDataSetChanged();
    }

    @Override
    public MovieAdapter.MovieAdapterViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.list_items, parent, false);
        return new MovieAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(MovieAdapter.MovieAdapterViewHolder holder, int position) {
        holder.bind(mMovies.get(position));
    }

    @Override
    public int getItemCount() {
        return mMovies.size();
    }

    public interface MovieAdapterOnClickHandler {
        void onClick(Movie m);
    }

    public class MovieAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public final ImageView mImageView;

        public MovieAdapterViewHolder(View view) {
            super(view);
            mImageView = view.findViewById(R.id.image_thumb);
            view.setOnClickListener(this);
        }

        void bind(Movie m) {
            Picasso p = Picasso.with(mImageView.getContext());
            p.load(m.getPoster()).placeholder(R.mipmap.ic_launcher).error(R.mipmap.not_found).into(mImageView);
        }

        @Override
        public void onClick(View view) {
            int adapterPosition = getAdapterPosition();
            Movie m = mMovies.get(adapterPosition);
            mOnClickHandler.onClick(m);
        }
    }
}

File: app/src/main/java/com/appsys/android/popularmovie/adapter/MovieAdapter.kt
package com.appsys.android.popularmovie.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import com.appsys.android.popularmovie.adapter.MovieAdapter.MovieAdapterViewHolder
import com.appsys.android.popularmovie.classes.Movie
import com.shakirfattani.course.movielisting.R
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.list_items.view.*
import java.util.*

/**
 * Created by shakir on 8/2/2017.
 */
class MovieAdapter(private val mOnClickHandler: MovieAdapterOnClickHandler, private val mMovies: ArrayList<Movie> = ArrayList()) : RecyclerView.Adapter<MovieAdapterViewHolder>() {
    fun resetMovieList() {
        mMovies.clear()
        notifyDataSetChanged()
    }

    var moviesData: ArrayList<Movie>
        get() = mMovies
        set(movieArray) {
            mMovies.addAll(movieArray)
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int)
            = MovieAdapterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.list_items, parent, false))

    override fun onBindViewHolder(holder: MovieAdapterViewHolder, position: Int)
        = holder.bind(mMovies[position])

    override fun getItemCount() = mMovies.size

    interface MovieAdapterOnClickHandler {
        fun onClick(m: Movie)
    }

    inner class MovieAdapterViewHolder(val view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {
        fun bind(m: Movie) {
            Picasso.with(view.context)
                    .load(m.poster)
                    .placeholder(R.mipmap.ic_launcher)
                    .error(R.mipmap.not_found)
                    .into(view.image_thumb)
        }

        override fun onClick(view: View) {
            mOnClickHandler.onClick(mMovies[adapterPosition])
        }

        init {
            view.setOnClickListener(this)
        }
    }

}